### DISCUSSION TITLE
[Desglose] Disparo

### DISCUSSION THREAD

**Discussion by emrodk** [Godot-en-Espanol/Node-City#9] [Tareas]
Primer ejercicio de desglose de tareas. Arrancamos por el disparo, una vez le sacamos la mano, hacemos lo mismo con el resto. A continuación vemos una descripción de disparo y como se comporta con el resto de los elementos de juego. Traten de pensar que otras cosas pueden afectar al disparo o si existe un comportamiento del mismo que no este descripto.
# Disparo
> El disparo es un objeto que se desplaza a velocidad constante y colisiona.

### Colisiona con...
- tanques:
  - no tiene armadura, lo destruye
  - tiene armadura, necesita 4 disparos para ser destruido, si es especial solo 2
- bloques:
  - si es de ladrillo, lo destruye
  - si es de metal, no lo destruye a menos que sea un disparo especial
- otro disparo, se destruyen ambos
- pared (Límite de pantalla) , solo se destruye el disparo
- Cualquier otro objeto es atravesado

### Atributos identificados:
- Velocidad porque se mueve
- Dirección de movimiento. Se mueve en 4 direcciones.
- Daño porque hay objetos que requieren más de un disparo para ser destruido y eso puede variar también (disparo especial)
- Tipo de daño porque las cosas que puede romper puede variar (ej romper metal)

### Señales:
- ~~cuando choca. (Esto es importante si hay un límite de balas en pantalla dependiendo del tanque).~~ Reemplazado por cadencia de disparo en tanque

### Relación con otros elementos de juego
- Tanque: Es instanciada por un tanque en su jerarquia o en la del padre
- Bloques: Algunos son atravesados con otros colisiona

**Comment by azagaya** [comment]
Que sería "pared" cuando decís "solo se destrulle la bala"? Pared de los límites de la pantalla?
Yo no se si en lugar de poner límite de balas en pantalla, no pondria un rounds per second, para limitar lo rápido que podes disparar

**Reply by emrodk** [reply]
Sería límite de pantalla, ahí lo agrego.

**Reply by emrodk** [reply]
Eso podría ser también, da más libertad al momento de disparar. Sería cosa del tanque regularlo

**Comment by LeandroCuello** [comment]
Deberíamos pensar dónde son instanciadas?
Como los tanques se van a mover y pueden ser liberados, las balas tendrías que ser instanciadas en un nodo que las contengan.

**Reply by emrodk** [reply]
es verdad, ahí lo anote!

**Reply by emrodk** [reply]
después cuando ampliemos tanque ya sabemos que hay que detallar esto

**Comment by LeandroCuello** [comment]
No se si hay una mejor forma, pero tal vez falte una variable que de la dirección en la que se mueve la bala. Luego que desde el código dónde se dispare este valor se modifiqué.

**Reply by emrodk** [reply]
bien ahí atento! jaja Si, necesita una dirección y el valor podría tomarla de la mira del tanque apenas se instancia

**Comment by emrodk** [comment]
@gfazzito Que efecto o sonido tendría el disparo? Cuando se mueve, cuando choca, cuando se dispara?

**Reply by gfazzito** [reply]
### Condiciones sonidos disparos
Viendo los modelos de gabo, el sonido del disparo está sumamente condicionado a:
- tamaño
- cantidad de cañones
- velocidad de disparo

Ejempo: Un cañon grande con tiempos de disparos largos debería tener un sonido pesado, lento y potente. Por el contrario, varios cañones chicos con disparos rápidos tendrán sonidos más agudos y livianos.
Estas variables se encuentran incluidas en la clasificación de enemigos (El tanque rápido tiene un cañon chico, un tanque pesado con 5 vidas tiene un cañon grande, etc).

### Sonidos de choque 
Dependerán del material contra que impacta (metal, ladrillo, otro disparo, etc)
- Colisión contra tanques: ver tipos de enemigos y salud de los mismos.
- Colisión contra pared: ver tipos de materiales
- Colisión contra otro disparo: ver tipos de disparo
- Colisión contra bandera: sonido único
- Colisión contra extremos del mapa: A definir

### Sonido de trayectoria del disparo
En principio no seria necesario. No provee de información necesaria y puede llegar a ser molesto si hay muchos disparos en simultáneo

### Disparos del jugador vs enemigos
Se podrían hacer los disparos enemigos con características sonoras similares que contrasten con las del jugador. Esto da mayor información sobre los disparos

### Disparos en movimiento
Son los mismos a disparar estando quieto
